File: ui-tests/PageObjects/Panes/Editor/EditorLine.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects.Components;

namespace UiTests.PageObjects.Panes.Editor;

/// <summary>
/// Represents a single rendered line inside an editor pane.
/// </summary>
public class EditorLine
{
    public EditorLine(EditorPane parentPane, ILocator root, int lineNumber)
    {
        ParentPane = parentPane;
        Root = root;
        LineNumber = lineNumber;
    }

    /// <summary>
    /// Parent editor pane hosting the line.
    /// </summary>
    public EditorPane ParentPane { get; }

    /// <summary>
    /// Locator representing the text portion of the line.
    /// </summary>
    public ILocator Root { get; }

    /// <summary>
    /// Line number as rendered by Monaco. A value of -1 indicates that the attribute was not available.
    /// </summary>
    public int LineNumber { get; }

    private void EnsureValidLineNumber()
    {
        if (LineNumber <= 0)
        {
            throw new InvalidOperationException("This editor line does not expose a valid line number.");
        }
    }

    /// <summary>
    /// Gutter element aligned with this line.
    /// </summary>
    public ILocator GutterElement()
    {
        EnsureValidLineNumber();
        return ParentPane.GutterElement(LineNumber);
    }

    /// <summary>
    /// Line number text within the gutter.
    /// </summary>
    public ILocator GutterLineNumberElement()
        => GutterElement().Locator(".gutter-line");

    /// <summary>
    /// Enabled tracepoint icon within the gutter.
    /// </summary>
    public ILocator GutterTraceIcon()
        => GutterElement().Locator(".gutter-trace");

    /// <summary>
    /// Disabled tracepoint icon within the gutter.
    /// </summary>
    public ILocator GutterDisabledTraceIcon()
        => GutterElement().Locator(".gutter-disabled-trace");

    /// <summary>
    /// Enabled breakpoint glyph inside the gutter.
    /// </summary>
    public ILocator GutterBreakpointEnabledIcon()
        => GutterElement().Locator(".gutter-breakpoint-enabled");

    /// <summary>
    /// Disabled breakpoint glyph inside the gutter.
    /// </summary>
    public ILocator GutterBreakpointDisabledIcon()
        => GutterElement().Locator(".gutter-breakpoint-disabled");

    /// <summary>
    /// Error state breakpoint glyph inside the gutter.
    /// </summary>
    public ILocator GutterBreakpointErrorIcon()
        => GutterElement().Locator(".gutter-breakpoint-error");

    /// <summary>
    /// Placeholder element indicating no breakpoint is set.
    /// </summary>
    public ILocator GutterNoBreakpointPlaceholder()
        => GutterElement().Locator(".gutter-no-breakpoint");

    /// <summary>
    /// Placeholder element indicating no tracepoint is set.
    /// </summary>
    public ILocator GutterNoTracePlaceholder()
        => GutterElement().Locator(".gutter-no-trace");

    /// <summary>
    /// Marker highlighting the current debugger location in the gutter.
    /// </summary>
    public ILocator GutterHighlightMarker()
        => GutterElement().Locator(".gutter-highlight-active");

    /// <summary>
    /// Omniscient flow values displayed in a loop context.
    /// </summary>
    public ILocator FlowLoopValueElements()
        => Root.Locator(".flow-loop-value");

    /// <summary>
    /// Omniscient flow values displayed for parallel executions.
    /// </summary>
    public ILocator FlowParallelValueElements()
        => Root.Locator(".flow-parallel-value");

    /// <summary>
    /// Variable name labels for loop flow values.
    /// </summary>
    public ILocator FlowLoopValueNameElements()
        => Root.Locator(".flow-loop-value-name");

    /// <summary>
    /// Variable name labels for parallel flow values.
    /// </summary>
    public ILocator FlowParallelValueNameElements()
        => Root.Locator(".flow-parallel-value-name");

    /// <summary>
    /// Editable expression field for omniscient loop values.
    /// </summary>
    public ILocator FlowLoopTextarea()
        => Root.Locator(".flow-loop-textarea");

    /// <summary>
    /// Multi-line value boxes rendered beside the code line.
    /// </summary>
    public ILocator FlowMultilineValueBoxes()
        => Root.Locator(".flow-multiline-value-box");

    /// <summary>
    /// Pointer element associated with multi-line value boxes.
    /// </summary>
    public ILocator FlowMultilineValuePointers()
        => Root.Locator(".flow-multiline-value-pointer");

    /// <summary>
    /// Returns all flow values currently attached to this line.
    /// </summary>
    /// <remarks>
    /// Note: Flow values are typically rendered as Monaco content widgets and may not
    /// be direct children of line elements. For comprehensive flow value discovery,
    /// consider using <see cref="EditorPane.FlowValuesAsync"/> instead.
    /// </remarks>
    public async Task<IReadOnlyList<FlowValue>> FlowValuesAsync()
    {
        var selector = ".flow-parallel-value-box, .flow-inline-value-box, .flow-loop-value-box, .flow-multiline-value-box";
        var locators = await Root.Locator(selector).AllAsync();
        var menu = new ContextMenu(ParentPane.Root.Page);
        var values = new List<FlowValue>();
        foreach (var locator in locators)
        {
            values.Add(new FlowValue(locator, menu));
        }

        return values;
    }

    /// <summary>
    /// Determines whether this line currently displays any breakpoint glyph.
    /// </summary>
    public async Task<bool> HasBreakpointAsync()
    {
        var breakpointLocator = GutterElement()
            .Locator(".gutter-breakpoint-enabled, .gutter-breakpoint-disabled, .gutter-breakpoint-error");
        return await breakpointLocator.CountAsync() > 0;
    }

    /// <summary>
    /// Determines whether this line currently has a tracepoint glyph.
    /// </summary>
    public async Task<bool> HasTracepointAsync()
    {
        var traceLocator = GutterElement().Locator(".gutter-trace, .gutter-disabled-trace");
        return await traceLocator.CountAsync() > 0;
    }

    /// <summary>
    /// Indicates whether the line hosts a tracepoint that is disabled.
    /// Returns <c>false</c> when no tracepoint is present.
    /// </summary>
    public async Task<bool> IsTracepointDisabledAsync()
    {
        if (!await HasTracepointAsync())
        {
            return false;
        }

        return await GutterDisabledTraceIcon().CountAsync() > 0;
    }

    /// <summary>
    /// Determines whether the line shows an error glyph in the gutter.
    /// </summary>
    public async Task<bool> HasErrorIconAsync()
        => await GutterBreakpointErrorIcon().CountAsync() > 0;

    /// <summary>
    /// Retrieves the rendered text for this line as shown by Monaco.
    /// </summary>
    public async Task<string> LineTextAsync()
    {
        var text = await Root.InnerTextAsync();
        return text?.Replace('\u00A0', ' ') ?? string.Empty;
    }

    /// <summary>
    /// Retrieves omniscient flow values rendered next to this line.
    /// </summary>
    public async Task<IReadOnlyList<string>> FlowParallelValuesAsync()
    {
        var values = new List<string>();
        var valueLocators = await FlowParallelValueElements().AllAsync();
        foreach (var locator in valueLocators)
        {
            var text = await locator.TextContentAsync();
            if (!string.IsNullOrWhiteSpace(text))
            {
                values.Add(text.Trim());
            }
        }

        return values;
    }
}

File: tsc-ui-tests/page-objects/panes/editor/editor-line.ts
import type { Locator, Page } from "@playwright/test";
import type { EditorPane } from "./editor-pane";
import { FlowValue } from "./flow-value";
import { ContextMenu } from "../../components/context-menu";

/**
 * Represents a single rendered line inside an editor pane.
 *
 * Port of ui-tests/PageObjects/Panes/Editor/EditorLine.cs
 */
export class EditorLine {
  readonly parentPane: EditorPane;
  readonly root: Locator;
  readonly lineNumber: number;

  constructor(parentPane: EditorPane, root: Locator, lineNumber: number) {
    this.parentPane = parentPane;
    this.root = root;
    this.lineNumber = lineNumber;
  }

  private ensureValidLineNumber(): void {
    if (this.lineNumber <= 0) {
      throw new Error("This editor line does not expose a valid line number.");
    }
  }

  gutterElement(): Locator {
    this.ensureValidLineNumber();
    return this.parentPane.gutterElement(this.lineNumber);
  }

  gutterLineNumberElement(): Locator {
    return this.gutterElement().locator(".gutter-line");
  }

  gutterTraceIcon(): Locator {
    return this.gutterElement().locator(".gutter-trace");
  }

  gutterDisabledTraceIcon(): Locator {
    return this.gutterElement().locator(".gutter-disabled-trace");
  }

  gutterBreakpointEnabledIcon(): Locator {
    return this.gutterElement().locator(".gutter-breakpoint-enabled");
  }

  gutterBreakpointDisabledIcon(): Locator {
    return this.gutterElement().locator(".gutter-breakpoint-disabled");
  }

  gutterBreakpointErrorIcon(): Locator {
    return this.gutterElement().locator(".gutter-breakpoint-error");
  }

  gutterNoBreakpointPlaceholder(): Locator {
    return this.gutterElement().locator(".gutter-no-breakpoint");
  }

  gutterNoTracePlaceholder(): Locator {
    return this.gutterElement().locator(".gutter-no-trace");
  }

  gutterHighlightMarker(): Locator {
    return this.gutterElement().locator(".gutter-highlight-active");
  }

  flowLoopValueElements(): Locator {
    return this.root.locator(".flow-loop-value");
  }

  flowParallelValueElements(): Locator {
    return this.root.locator(".flow-parallel-value");
  }

  flowLoopValueNameElements(): Locator {
    return this.root.locator(".flow-loop-value-name");
  }

  flowParallelValueNameElements(): Locator {
    return this.root.locator(".flow-parallel-value-name");
  }

  flowLoopTextarea(): Locator {
    return this.root.locator(".flow-loop-textarea");
  }

  flowMultilineValueBoxes(): Locator {
    return this.root.locator(".flow-multiline-value-box");
  }

  flowMultilineValuePointers(): Locator {
    return this.root.locator(".flow-multiline-value-pointer");
  }

  async flowValues(): Promise<FlowValue[]> {
    const selector =
      ".flow-parallel-value-box, .flow-inline-value-box, .flow-loop-value-box, .flow-multiline-value-box";
    const locators = await this.root.locator(selector).all();
    const menu = new ContextMenu(this.parentPane.root.page()!);
    return locators.map((l) => new FlowValue(l, menu));
  }

  async hasBreakpoint(): Promise<boolean> {
    const locator = this.gutterElement().locator(
      ".gutter-breakpoint-enabled, .gutter-breakpoint-disabled, .gutter-breakpoint-error",
    );
    return (await locator.count()) > 0;
  }

  async hasTracepoint(): Promise<boolean> {
    const locator = this.gutterElement().locator(".gutter-trace, .gutter-disabled-trace");
    return (await locator.count()) > 0;
  }

  async isTracepointDisabled(): Promise<boolean> {
    if (!(await this.hasTracepoint())) {
      return false;
    }
    return (await this.gutterDisabledTraceIcon().count()) > 0;
  }

  async hasErrorIcon(): Promise<boolean> {
    return (await this.gutterBreakpointErrorIcon().count()) > 0;
  }

  async lineText(): Promise<string> {
    const text = await this.root.innerText();
    return text.replace(/\u00A0/g, " ");
  }

  async flowParallelValues(): Promise<string[]> {
    const values: string[] = [];
    const locators = await this.flowParallelValueElements().all();
    for (const locator of locators) {
      const text = await locator.textContent();
      if (text && text.trim().length > 0) {
        values.push(text.trim());
      }
    }
    return values;
  }
}
